### PR TITLE
add storage VLAN interface on all slurm nodes

### PR DESCRIPTION
**Pull request by scrungus**
Additional note from @sjpb :
> The `.caas` environment uses `etc_hosts` to template out /etc/hosts. By default (= .caas environment) it uses `ansible_host` as the IP. This is safe after this IP as `stackhpc.terraform.infra/inventory_adopt` uses the terraform output `node.ip` for this, and this is set in `roles/cluster_infra/templates/outputs.tf` to be network[0] on all nodes. This PR adds the storage network as the *2nd* network so it will not be in /etc/hosts.

### CODE REVIEW

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 1]
```diff
@@ -224,7 +257,29 @@ resource "openstack_networking_port_v2" "control" {
...
+  admin_state_up = "true"
+
+  # not sure if needed here
+  security_group_ids = [
+    "${openstack_networking_secgroup_v2.secgroup_slurm_cluster.id}"
+  ]
+
+  # for now, until we set up flavors
+  binding {
+    vnic_type = "normal"
```
Comment (JohnGarbutt) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:274]: lets add cluster_vnic_type a bit like the existing one:

`vnic_type = "{{ cluster_vnic_type | default('normal') }}"`

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 2]
```diff
@@ -116,6 +116,11 @@ data "openstack_networking_network_v2" "cluster_external_network" {
   name = "{{ cluster_external_network }}"
 }
 
+# Always get the SRIOV storage network 
+data "openstack_networking_network_v2" "portal_storage_direct" {
+  tags = ["portal-storage-direct"]
```
Comment (JohnGarbutt) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:121]: Lets make this something like cluster_storage_network, and we skip attaching to it if its not defined via extra vars?

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 3]
```diff
@@ -201,6 +211,29 @@ resource "openstack_networking_port_v2" "login" {
...
 
+# Storage VLAN
+resource "openstack_networking_port_v2" "login_portal_storage" {
+  name           = "{{ cluster_name }}-login-0-portal-storage"
+  network_id     = data.openstack_networking_network_v2.portal_storage_direct.id
+  admin_state_up = "true"
+
+  # not sure if needed here
+  security_group_ids = [
+    "${openstack_networking_secgroup_v2.secgroup_slurm_cluster.id}",
```
[the remark this answers is not in the document]
Comment (sjpb) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:222]: If we're on OVN we need stateless secgroups to get decent performance. Not sure if that's relevant here?

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 4]
```diff
@@ -201,6 +211,29 @@ resource "openstack_networking_port_v2" "login" {
   }
 }
 
+# Storage VLAN
+resource "openstack_networking_port_v2" "login_portal_storage" {
+  name           = "{{ cluster_name }}-login-0-portal-storage"
```
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:216]: what's the `-portal-` bit of the name here?

File: ansible/roles/cluster_infra/templates/resources.tf.j2 (hunk 1 above)
Comment (sjpb) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:274]: Same for other port. And in fact it'd be nice to make it clear that this doesn't apply to the non-storage ports, so something like `cluster_storage_vnic_type`.

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 5]
```diff
@@ -224,7 +257,29 @@ resource "openstack_networking_port_v2" "control" {
   }
 }
 
+# Storage VLAN
+
+resource "openstack_networking_port_v2" "control_portal_storage" {
+  name           = "{{ cluster_name }}-control-0-portal-storage"
+  network_id     = data.openstack_networking_network_v2.portal_storage_direct.id
```
Comment (sjpb) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:264]: again why `portal`? `storage` I think makes some sense c.f. something more generic, implies it probably does not provide external connectivity, and hopefully high perf.

**Comment by scrungus** [comment]
`cluster_storage_network` changes here https://github.com/stackhpc/azimuth/pull/236

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 6]
```diff
@@ -201,6 +213,30 @@ resource "openstack_networking_port_v2" "login" {
...
+  name           = "{{ cluster_name }}-login-storage-0"
+  network_id     = data.openstack_networking_network_v2.cluster_storage_direct.id
+  admin_state_up = "true"
+
+  # not sure if needed here
+  security_group_ids = [
+    "${openstack_networking_secgroup_v2.secgroup_slurm_cluster.id}",
+  ]
+
+  # for now, until we set up flavors
```
Comment (JohnGarbutt) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:228]: lets remove the comment here, I think what you have is the end state anyway.

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 7]
```diff
@@ -116,6 +116,13 @@ data "openstack_networking_network_v2" "cluster_external_network" {
   name = "{{ cluster_external_network }}"
 }
 
+# SRIOV storage network 
+{% if cluster_storage_network is defined %}
+data "openstack_networking_network_v2" "cluster_storage_direct" {
```
Comment (JohnGarbutt) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:121]: lets just call this cluster_storage for now, so we match the cluster_storage_network parameter please?

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 8]
```diff
@@ -201,6 +213,30 @@ resource "openstack_networking_port_v2" "login" {
...
 }
 
+# Storage VLAN
+{% if cluster_storage_network is defined %}
+resource "openstack_networking_port_v2" "login_storage" {
+  name           = "{{ cluster_name }}-login-storage-0"
+  network_id     = data.openstack_networking_network_v2.cluster_storage_direct.id
+  admin_state_up = "true"
+
+  # not sure if needed here
```
Comment (JohnGarbutt) [inline ansible/roles/cluster_infra/templates/resources.tf.j2:223]: I think we need it, for the moment.

We might have to eventually make this option, when we test with VF-LAG, etc.

**Review by JohnGarbutt — CHANGES_REQUESTED** [review]
Nice, its looking good.

Mostly just a few nits around removing the comments we don't need, and renaming the network to be called "storage" rather than "storage_direct", just so we are consistent with the param.

**Comment by scrungus** [comment]
`cluster_vnic_profile` also removed here from the primary network as it is no longer used

**Review by JohnGarbutt — APPROVED** [review]
I think this looks good now. @sjpb are you happy with this too?

**Comment by sjpb** [comment]
> `cluster_vnic_profile` also removed here from the primary network as it is no longer used

@scrungus why not?

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 9]
```diff
@@ -193,14 +205,31 @@ resource "openstack_networking_port_v2" "login" {
 
   binding {
     vnic_type = "{{ cluster_vnic_type | default('normal') }}"
-    {% if cluster_vnic_profile is defined %}
```
[the remark this answers is not in the document]
Comment (JohnGarbutt) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:196]: That was from me, we shouldn't ever use vnic_profile any more, its fixed in yoga.

File: ansible/roles/cluster_infra/templates/resources.tf.j2 [hunk 10]
```diff
@@ -193,14 +205,31 @@ resource "openstack_networking_port_v2" "login" {
...
+  name           = "{{ cluster_name }}-login-storage-0"
+  network_id     = data.openstack_networking_network_v2.cluster_storage.id
+  admin_state_up = "true"
+
+  security_group_ids = [
+    "${openstack_networking_secgroup_v2.secgroup_slurm_cluster.id}",
+  ]
+
+  binding {
+    vnic_type = "{{ cluster_storage_vnic_type | default('normal') }}"
```
[the remark this answers is not in the document]
Comment (JohnGarbutt) [reply ansible/roles/cluster_infra/templates/resources.tf.j2:223]: So this one is specific to the storage network. We need normal ports on the cluster network and direct mode ports on the storage network. Not sure this has been tested yet though.

**Pull request merged**